In [5]:
# remove hash to install package
# !pip3 install owslib==0.25.0 fiona==1.8.21 geopandas==0.10.2 requests==2.28.0 folium==0.12.1

In [2]:
from owslib.wfs import WebFeatureService
import geopandas
import folium
import io
import zipfile
import pandas as pd
import os
from urllib.request import urlretrieve

import warnings
warnings.filterwarnings("ignore")

## Generate POA to SA2 Lookup Table

In [3]:
# load selected external data
pop_df = geopandas.read_file(f'../data/abs/2021_population_census.gml')
sa2_bound = geopandas.read_file(f'../data/abs/sa2_boundaries.gml')
poa_bound = geopandas.read_file(f'../data/abs/poa_boundaries.gml')
sa2_income = geopandas.read_file(f'../data/abs/sa2_income.gml')

# Open zipfile
unzip_poa_sa2 = zipfile.ZipFile('../data/abs/poa_sa2_lookup.zip')
poa_to_sa2 = pd.read_excel(unzip_poa_sa2
                           .open('1270055006_CG_POSTCODE_2011_SA2_2011.xls')
                        , sheet_name='Table 3', skiprows=5)

In [4]:
# Select 2021 population census
area_id = ['gml_id', 'primaryindex', 'state_code_2016', 'sa2_maincode_2016',
           'sa2_name_2016']
col_2021 = [x for x in pop_df.columns if '2021' in x or '2020_21' in x]
pop_21 = pop_df[area_id + col_2021]

In [5]:
pop_21.shape

(2292, 18)

There should be 2,310 SA2 regions.

In [6]:
# ERP refers to estimated resident population
pop_21.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2292 entries, 0 to 2291
Data columns (total 18 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   gml_id                           2292 non-null   object 
 1   primaryindex                     2292 non-null   int64  
 2   state_code_2016                  2292 non-null   int64  
 3   sa2_maincode_2016                2292 non-null   int64  
 4   sa2_name_2016                    2292 non-null   object 
 5   erp_2021                         2292 non-null   int64  
 6   erp_change_number_2020_21        2292 non-null   int64  
 7   erp_change_per_cent_2020_21      2292 non-null   float64
 8   pop_density_2021_people_per_km2  2292 non-null   float64
 9   births_2020_21                   2288 non-null   float64
 10  deaths_2020_21                   2288 non-null   float64
 11  natural_increase_2020_21         2288 non-null   float64
 12  internal_arrivals_20

Do we want to include these regions in the data???
If so, will we include the attributes with missing data?
If yes, how?
If no, why?

In [7]:
pop_21[pop_21["births_2020_21"].isnull()]

gml_id  primaryindex  \
2231  abs_regional_population_sa2_2001_2021.2292          2292   
2289  abs_regional_population_sa2_2001_2021.2289          2289   
2290  abs_regional_population_sa2_2001_2021.2290          2290   
2291  abs_regional_population_sa2_2001_2021.2291          2291   

      state_code_2016  sa2_maincode_2016            sa2_name_2016  erp_2021  \
2231                9          901041004           Norfolk Island      1749   
2289                9          901011001         Christmas Island      1979   
2290                9          901021002  Cocos (Keeling) Islands       579   
2291                9          901031003               Jervis Bay       397   

      erp_change_number_2020_21  erp_change_per_cent_2020_21  \
2231                         14                       0.8069   
2289                         15                       0.7637   
2290                          6                       1.0471   
2291                          2                       0.5063   

      pop_density_2021_people_per_km2  births_2020_21  deaths_2020_21  \
2231                        45.251202             NaN             NaN   
2289                        14.537000             NaN             NaN   
2290                        42.212601             NaN             NaN   
2291                         5.854300             NaN             NaN   

      natural_increase_2020_21  internal_arrivals_2020_21  \
2231                       NaN                        NaN   
2289                       NaN                        NaN   
2290                       NaN                        NaN   
2291                       NaN                        NaN   

      internal_departures_2020_21  net_internal_migration_2020_21  \
2231                          NaN                             NaN   
2289                          NaN                             NaN   
2290                          NaN                             NaN   
2291                          NaN                             NaN   

      overseas_arrivals_2020_21  overseas_departures_2020_21  \
2231                        NaN                          NaN   
2289                        NaN                          NaN   
2290                        NaN                          NaN   
2291                        NaN                          NaN   

      net_overseas_migration_2020_21  
2231                             NaN  
2289                             NaN  
2290                             NaN  
2291                             NaN

In [8]:
poa_to_sa2 = poa_to_sa2.dropna()

There are 2162 SA2 codes in Postcode to SA2 data, but there are 2292. Try to find newer data for poa_to_sa2?

In [9]:
poa_to_sa2["SA2_MAINCODE_2011"].unique().size

2162

In [10]:
poa_to_sa2

POSTCODE  POSTCODE.1  SA2_MAINCODE_2011      SA2_NAME_2011     RATIO  \
1        0800       800.0        701011002.0        Darwin City  1.000000   
2        0810       810.0        701021010.0              Alawa  0.071997   
3        0810       810.0        701021013.0   Brinkin - Nakara  0.096392   
4        0810       810.0        701021016.0      Coconut Grove  0.096494   
5        0810       810.0        701021018.0            Jingili  0.061562   
...       ...         ...                ...                ...       ...   
5984     7466      7466.0        604031097.0  West Coast (Tas.)  1.000000   
5985     7467      7467.0        604031097.0  West Coast (Tas.)  1.000000   
5986     7468      7468.0        604031097.0  West Coast (Tas.)  1.000000   
5987     7469      7469.0        604031097.0  West Coast (Tas.)  1.000000   
5988     7470      7470.0        604031097.0  West Coast (Tas.)  1.000000   

      PERCENTAGE  
1      99.999998  
2       7.199707  
3       9.639178  
4       9.649355  
5       6.156198  
...          ...  
5984  100.000000  
5985  100.000000  
5986  100.000000  
5987  100.000000  
5988  100.000000  

[5988 rows x 6 columns]

In [11]:
poa_bound

gml_id  primaryindex  objectid  poa_code_2016 poa_name_2016  \
0        poa_2016_aust.1             1         1            800          0800   
1        poa_2016_aust.2             2         2            810          0810   
2        poa_2016_aust.3             3         3            812          0812   
3        poa_2016_aust.4             4         4            815          0815   
4        poa_2016_aust.5             5         5            820          0820   
...                  ...           ...       ...            ...           ...   
2665  poa_2016_aust.2639          2639      2639           7268          7268   
2666  poa_2016_aust.2640          2640      2640           7270          7270   
2667  poa_2016_aust.2641          2641      2641           7275          7275   
2668  poa_2016_aust.2642          2642      2642           7276          7276   
2669  poa_2016_aust.2643          2643      2643           7277          7277   

      area_albers_sqkm                                           geometry  
0               3.1734  POLYGON ((130.83450 -12.45800, 130.83390 -12.4...  
1              23.7902  POLYGON ((130.84710 -12.37750, 130.84730 -12.3...  
2              35.8899  POLYGON ((130.89190 -12.36880, 130.89220 -12.3...  
3               0.6381  POLYGON ((130.87240 -12.37650, 130.87230 -12.3...  
4              39.0462  POLYGON ((130.83500 -12.43010, 130.83510 -12.4...  
...                ...                                                ...  
2665          162.4060  POLYGON ((147.20120 -41.25830, 147.20130 -41.2...  
2666          265.1620  MULTIPOLYGON (((146.82300 -41.13960, 146.82290...  
2667          340.3100  POLYGON ((146.84830 -41.24130, 146.84850 -41.2...  
2668            5.9926  POLYGON ((146.97820 -41.28340, 146.97820 -41.2...  
2669           98.6653  MULTIPOLYGON (((147.08800 -41.38310, 147.08800...  

[2670 rows x 7 columns]

In [12]:
sa2_bound

gml_id  primaryindex  sa2_maincode_2016  \
0        sa2_2016_aust.1             1          101021007   
1        sa2_2016_aust.2             2          101021008   
2        sa2_2016_aust.9             9          101031015   
3       sa2_2016_aust.10            10          101031016   
4       sa2_2016_aust.11            11          101041017   
...                  ...           ...                ...   
2305  sa2_2016_aust.2305          2305          901011001   
2306  sa2_2016_aust.2306          2306          901021002   
2307  sa2_2016_aust.2307          2307          901031003   
2308  sa2_2016_aust.2308          2308          901041004   
2309  sa2_2016_aust.2309          2309          997979799   

      sa2_5digitcode_2016                         sa2_name_2016  \
0                   11007                             Braidwood   
1                   11008                               Karabar   
2                   11015                          Cooma Region   
3                   11016                 Jindabyne - Berridale   
4                   11017                          Batemans Bay   
...                   ...                                   ...   
2305                91001                      Christmas Island   
2306                91002               Cocos (Keeling) Islands   
2307                91003                            Jervis Bay   
2308                91004                        Norfolk Island   
2309                99799  Migratory - Offshore - Shipping (OT)   

      sa3_code_2016                         sa3_name_2016  sa4_code_2016  \
0             10102                            Queanbeyan            101   
1             10102                            Queanbeyan            101   
2             10103                       Snowy Mountains            101   
3             10103                       Snowy Mountains            101   
4             10104                           South Coast            101   
...             ...                                   ...            ...   
2305          90101                      Christmas Island            901   
2306          90102               Cocos (Keeling) Islands            901   
2307          90103                            Jervis Bay            901   
2308          90104                        Norfolk Island            901   
2309          99797  Migratory - Offshore - Shipping (OT)            997   

                             sa4_name_2016 gccsa_code_2016  \
0                           Capital Region           1RNSW   
1                           Capital Region           1RNSW   
2                           Capital Region           1RNSW   
3                           Capital Region           1RNSW   
4                           Capital Region           1RNSW   
...                                    ...             ...   
2305                     Other Territories           9OTER   
2306                     Other Territories           9OTER   
2307                     Other Territories           9OTER   
2308                     Other Territories           9OTER   
2309  Migratory - Offshore - Shipping (OT)           99799   

                           gccsa_name_2016  state_code_2016  \
0                              Rest of NSW                1   
1                              Rest of NSW                1   
2                              Rest of NSW                1   
3                              Rest of NSW                1   
4                              Rest of NSW                1   
...                                    ...              ...   
2305                     Other Territories                9   
2306                     Other Territories                9   
2307                     Other Territories                9   
2308                     Other Territories                9   
2309  Migratory - Offshore - Shipping (OT)                9   

        state_name_2016  area_albers_sqkm  \
0       New South Wa

In [13]:
poa_bound["geometry"] = poa_bound["geometry"].to_crs(epsg=4326)
sa2_bound["geometry"] = sa2_bound["geometry"].to_crs(epsg=4326)

In [14]:
poa_bound["centroid"] = poa_bound["geometry"].centroid

/var/folders/kn/0h56wsl91tz9kbw4pcdll7zw0000gn/T/ipykernel_4017/1735245261.py:1: UserWarning: Geometry is in a geographic CRS. Results from 'centroid' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation.

  poa_bound["centroid"] = poa_bound["geometry"].centroid


In [15]:
poa_bound = poa_bound[poa_bound["centroid"].notnull()]

In [16]:
sa2_bound = sa2_bound[sa2_bound["geometry"].notnull()]

In [17]:
from shapely.validation import make_valid
sa2_bound["geometry"] = sa2_bound["geometry"].apply(make_valid)

/Users/patrick/opt/anaconda3/lib/python3.9/site-packages/geopandas/geodataframe.py:1351: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  super().__setitem__(key, value)


In [18]:
from shapely.geometry import Point
poa_list = poa_bound["centroid"].to_list()
sa2_list = []
for point in poa_list:
    contains = sa2_bound["geometry"].contains(Point(point))
    sa2_list.append(contains.index[contains])

In [19]:
for i in range(len(sa2_list)):
    try:
        sa2_list[i] = sa2_list[i][0]
    except IndexError:
        sa2_list[i] = None
        print(i)

38
66
138
206
335
347
349
352
1408
1539
1784
1863
2160
2302
2492
2496
2524
2528
2531
2567
2603
2626


In [20]:
poa_bound["sa2_area"] = pd.Series(sa2_list)

In [21]:
poa_bound[poa_bound["sa2_area"].isnull()]

gml_id  primaryindex  objectid  poa_code_2016 poa_name_2016  \
38      poa_2016_aust.33            33        33            880          0880   
66      poa_2016_aust.55            55        55           2027          2027   
138    poa_2016_aust.122           122       122           2105          2105   
206   poa_2016_aust.2110          2110      2110           5575          5575   
335    poa_2016_aust.231           231       231           2230          2230   
347    poa_2016_aust.243           243       243           2261          2261   
349    poa_2016_aust.245           245       245           2263          2263   
352    poa_2016_aust.248           248       248           2267          2267   
1408  poa_2016_aust.1359          1359      1359           3960          3960   
1539  poa_2016_aust.1480          1480      1480           4178          4178   
1784  poa_2016_aust.1761          1761      1761           4802          4802   
1863  poa_2016_aust.1805          1805      1805           4875          4875   
2160  poa_2016_aust.2123          2123      2123           5606          5606   
2302  poa_2016_aust.2279          2279      2279           6168          6168   
2492  poa_2016_aust.2482          2482      2482           6530          6530   
2496  poa_2016_aust.2486          2486      2486           6537          6537   
2524  poa_2016_aust.2525          2525      2525           6707          6707   
2528  poa_2016_aust.2529          2529      2529           6713          6713   
2531  poa_2016_aust.2554          2554      2554           7001          7001   
2567  poa_2016_aust.2591          2591      2591           7150          7150   
2603  poa_2016_aust.2552          2552      2552           6799          6799   
2626  poa_2016_aust.2600          2600      2600           7175          7175   
2668  poa_2016_aust.2642          2642      2642           7276          7276   
2669  poa_2016_aust.2643          2643      2643           7277          7277   

      area_albers_sqkm                                           geometry  \
38             88.9707  MULTIPOLYGON (((136.67621 -12.23769, 136.67551...   
66              1.3761  MULTIPOLYGON (((151.25861 -33.85969, 151.25861...   
138             2.1888  MULTIPOLYGON (((151.28001 -33.63819, 151.28001...   
206          1367.3200  MULTIPOLYGON (((136.92601 -35.02879, 136.92601...   
335             9.1909  MULTIPOLYGON (((151.12451 -34.07999, 151.12451...   
347            51.4591  MULTIPOLYGON (((151.50201 -33.33019, 151.50201...   
349            22.3942  MULTIPOLYGON (((151.57271 -33.29069, 151.57261...   
352             3.6556  POLYGON ((151.57021 -33.06989, 151.57021 -33.0...   
1408          831.5160  MULTIPOLYGON (((146.30501 -39.15189, 146.30471...   
1539           44.3278  MULTIPOLYGON (((153.18851 -27.40099, 153.18851...   
1784          276.4340  MULTIPOLYGON (((148.81011 -20.30289, 148.81001...   
1863          851.2620  MULTIPOLYGON (((142.66891 -10.66699, 142.66881...   
2160           88.5758  MULTIPOLYGON (((136.21141 -34.81539, 136.21121...   
2302           49.0225  MULTIPOLYGON (((115.66571 -32.12589, 115.66561...   
2492          127.2790  MULTIPOLYGON (((113.85901 -28.68149, 113.85881...   
2496         4454.1600  MULTIPOLYGON (((113.20091 -26.00139, 113.20021...   
2524         5978.8300  MULTIPOLYGON (((114.38061 -22.39379, 114.38011...   
2528          195.7720  MULTIPOLYGON (((116.62421 -20.66689, 116.62391...   
2531            2.9842  MULTIPOLYGON (((146.24680 -43.66159, 146.24680...   
2567          503.6590  MULTIPOLYGON (((147.42540 -43.25409, 147.42530...   
2603           13.7163  MULTIPOLYGON (((96.83051 -12.17639, 96.83051 -...   
2626           62.3831  POLYGON ((147.84560 -42.83369, 147.84470 -42.8...   
2668            5.9926  POLYGON ((146.97821 -41.28339, 146.97821 -41.2...   
2669           98.6653  MULTIPOLYGON (((147.08801 -41.38309, 147.08801...   

                         centroid  sa2_area  
38    POINT (136.646

In [22]:
poa_w_sa2 = poa_bound.merge(sa2_bound[['sa2_maincode_2016', 'sa2_name_2016','geometry']], how="left", left_on="sa2_area", right_on=sa2_bound.index)

In [23]:
poa_w_sa2

gml_id  primaryindex  objectid  poa_code_2016 poa_name_2016  \
0        poa_2016_aust.1             1         1            800          0800   
1        poa_2016_aust.2             2         2            810          0810   
2        poa_2016_aust.3             3         3            812          0812   
3        poa_2016_aust.4             4         4            815          0815   
4        poa_2016_aust.5             5         5            820          0820   
...                  ...           ...       ...            ...           ...   
2663  poa_2016_aust.2639          2639      2639           7268          7268   
2664  poa_2016_aust.2640          2640      2640           7270          7270   
2665  poa_2016_aust.2641          2641      2641           7275          7275   
2666  poa_2016_aust.2642          2642      2642           7276          7276   
2667  poa_2016_aust.2643          2643      2643           7277          7277   

      area_albers_sqkm                                         geometry_x  \
0               3.1734  POLYGON ((130.83451 -12.45799, 130.83391 -12.4...   
1              23.7902  POLYGON ((130.84711 -12.37749, 130.84731 -12.3...   
2              35.8899  POLYGON ((130.89191 -12.36879, 130.89221 -12.3...   
3               0.6381  POLYGON ((130.87241 -12.37649, 130.87231 -12.3...   
4              39.0462  POLYGON ((130.83501 -12.43009, 130.83511 -12.4...   
...                ...                                                ...   
2663          162.4060  POLYGON ((147.20121 -41.25829, 147.20131 -41.2...   
2664          265.1620  MULTIPOLYGON (((146.82301 -41.13959, 146.82291...   
2665          340.3100  POLYGON ((146.84831 -41.24129, 146.84851 -41.2...   
2666            5.9926  POLYGON ((146.97821 -41.28339, 146.97821 -41.2...   
2667           98.6653  MULTIPOLYGON (((147.08801 -41.38309, 147.08801...   

                         centroid  sa2_area  sa2_maincode_2016  \
0     POINT (130.84373 -12.46185)    2130.0        701011002.0   
1     POINT (130.87685 -12.36972)    1401.0        701021013.0   
2     POINT (130.91450 -12.37765)    1454.0        701021014.0   
3     POINT (130.86899 -12.37174)    1401.0        701021013.0   
4     POINT (130.86497 -12.42338)    2139.0        701011006.0   
...                           ...       ...                ...   
2663  POINT (147.21952 -41.29213)    2135.0        602021055.0   
2664  POINT (146.78669 -41.17828)    2135.0        602021055.0   
2665  POINT (146.81144 -41.31036)    2135.0        602021055.0   
2666  POINT (146.96355 -41.28260)       NaN                NaN   
2667  POINT (146.99725 -41.37181)       NaN                NaN   

               sa2_name_2016  \
0                Darwin City   
1           Brinkin - Nakara   
2              Buffalo Creek   
3           Brinkin - Nakara   
4     Ludmilla - The Narrows   
...                      ...   
2663    Grindelwald - Lanena   
2664    Grindelwald - Lanena   
2665    Grindelwald - Lanena   
2666                     NaN   
2667                     NaN   

                                             geometry_y  
0     POLYGON ((130.83451 -12.45799, 130.83391 -12.4...  
1     POLYGON ((130.86381 -12.36689, 130.86631 -12.3...  
2     POLYGON ((130.90101 -12.36579, 130.90081 -12.3...  
3     POLYGON ((130.86381 -12.36689, 130.86631 -12.3...  
4     POLYGON ((130.84491 -12.41539, 130.84511 -12.4...  
...                                                 ...  
2663  GEOMETRYCOLLECTION (POLYGON ((146.80531 -41.30...  
2664  GEOMETRYCOLLECTION (POLYGON ((146.80531 -41.30...  
2665  GEOMETRYCOLLECTION (POLYGON ((146.80531 -41.30...  
2666                                               None  
2667                                               None  

[2668 rows x 12 columns]

In [24]:
poa_w_sa2 = poa_w_sa2[["poa_code_2016", "poa_name_2016", "sa2_maincode_2016", "sa2_name_2016", "geometry_y"]]

In [25]:
poa_w_sa2

poa_code_2016 poa_name_2016  sa2_maincode_2016           sa2_name_2016  \
0               800          0800        701011002.0             Darwin City   
1               810          0810        701021013.0        Brinkin - Nakara   
2               812          0812        701021014.0           Buffalo Creek   
3               815          0815        701021013.0        Brinkin - Nakara   
4               820          0820        701011006.0  Ludmilla - The Narrows   
...             ...           ...                ...                     ...   
2663           7268          7268        602021055.0    Grindelwald - Lanena   
2664           7270          7270        602021055.0    Grindelwald - Lanena   
2665           7275          7275        602021055.0    Grindelwald - Lanena   
2666           7276          7276                NaN                     NaN   
2667           7277          7277                NaN                     NaN   

                                             geometry_y  
0     POLYGON ((130.83451 -12.45799, 130.83391 -12.4...  
1     POLYGON ((130.86381 -12.36689, 130.86631 -12.3...  
2     POLYGON ((130.90101 -12.36579, 130.90081 -12.3...  
3     POLYGON ((130.86381 -12.36689, 130.86631 -12.3...  
4     POLYGON ((130.84491 -12.41539, 130.84511 -12.4...  
...                                                 ...  
2663  GEOMETRYCOLLECTION (POLYGON ((146.80531 -41.30...  
2664  GEOMETRYCOLLECTION (POLYGON ((146.80531 -41.30...  
2665  GEOMETRYCOLLECTION (POLYGON ((146.80531 -41.30...  
2666                                               None  
2667                                               None  

[2668 rows x 5 columns]

In [26]:
# check the postcodes not allocated

undef_poa_ind = poa_w_sa2[poa_w_sa2["sa2_maincode_2016"].isnull()].index.values

In [27]:
undef_poa_ind

array([  38,   66,  138,  206,  335,  347,  349,  352, 1408, 1539, 1784,
       1863, 2160, 2302, 2492, 2496, 2524, 2528, 2531, 2567, 2603, 2624,
       2666, 2667])

In [28]:
df_undef_poa = (poa_bound[poa_bound.index.isin(undef_poa_ind)]
                [['poa_code_2016', 'poa_name_2016', 'sa2_area']])

In [29]:
df_undef_poa.head()

poa_code_2016 poa_name_2016  sa2_area
38             880          0880       NaN
66            2027          2027       NaN
138           2105          2105       NaN
206           5575          5575       NaN
335           2230          2230       NaN

In [30]:
# there are postcodes with multiple SA2s
display(poa_to_sa2.sort_values(by=['POSTCODE', 'PERCENTAGE'], ascending=True).head(20))

POSTCODE  POSTCODE.1  SA2_MAINCODE_2011     SA2_NAME_2011     RATIO  \
1      0800       800.0        701011002.0       Darwin City  1.000000   
13     0810       810.0        701021029.0           Wanguri  0.060207   
5      0810       810.0        701021018.0           Jingili  0.061562   
8      0810       810.0        701021024.0              Moil  0.068575   
6      0810       810.0        701021021.0        Lyons (NT)  0.070201   
12     0810       810.0        701021028.0           Wagaman  0.071421   
2      0810       810.0        701021010.0             Alawa  0.071997   
7      0810       810.0        701021023.0           Millner  0.083483   
11     0810       810.0        701021027.0              Tiwi  0.085041   
3      0810       810.0        701021013.0  Brinkin - Nakara  0.096392   
4      0810       810.0        701021016.0     Coconut Grove  0.096494   
10     0810       810.0        701021026.0       Rapid Creek  0.110723   
9      0810       810.0        701021025.0        Nightcliff  0.123904   
15     0812       812.0        701021014.0     Buffalo Creek  0.000162   
14     0812       812.0        701021011.0             Anula  0.124460   
19     0812       812.0        701021030.0            Wulagi  0.135592   
17     0812       812.0        701021020.0           Leanyer  0.242811   
18     0812       812.0        701021022.0   Malak - Marrara  0.246074   
16     0812       812.0        701021019.0            Karama  0.250887   
29     0820       820.0        701021015.0    Charles Darwin  0.000160   

    PERCENTAGE  
1    99.999998  
13    6.020652  
5     6.156198  
8     6.857527  
6     7.020120  
12    7.142110  
2     7.199707  
7     8.348290  
11    8.504137  
3     9.639178  
4     9.649355  
10   11.072343  
9    12.390383  
15    0.016237  
14   12.446020  
19   13.559184  
17   24.281085  
18   24.607349  
16   25.088725  
29    0.015946

In [31]:
# only get the SA2 with highest percentage
idx = poa_to_sa2.groupby(['POSTCODE'])['PERCENTAGE'].transform(max) == poa_to_sa2['PERCENTAGE']
high_perc_sa2_poa = poa_to_sa2[idx].reset_index(drop=True)

In [32]:
poa_w_sa2 = pd.merge(poa_w_sa2, high_perc_sa2_poa[['POSTCODE', 'SA2_MAINCODE_2011', 'SA2_NAME_2011']], how='left', left_on='poa_name_2016', right_on='POSTCODE')

In [34]:
poa_w_sa2

poa_code_2016 poa_name_2016  sa2_maincode_2016           sa2_name_2016  \
0               800          0800        701011002.0             Darwin City   
1               810          0810        701021013.0        Brinkin - Nakara   
2               812          0812        701021014.0           Buffalo Creek   
3               815          0815        701021013.0        Brinkin - Nakara   
4               820          0820        701011006.0  Ludmilla - The Narrows   
...             ...           ...                ...                     ...   
2664           7268          7268        602021055.0    Grindelwald - Lanena   
2665           7270          7270        602021055.0    Grindelwald - Lanena   
2666           7275          7275        602021055.0    Grindelwald - Lanena   
2667           7276          7276                NaN                     NaN   
2668           7277          7277                NaN                     NaN   

                                             geometry_y POSTCODE  \
0     POLYGON ((130.83451 -12.45799, 130.83391 -12.4...     0800   
1     POLYGON ((130.86381 -12.36689, 130.86631 -12.3...     0810   
2     POLYGON ((130.90101 -12.36579, 130.90081 -12.3...     0812   
3     POLYGON ((130.86381 -12.36689, 130.86631 -12.3...      NaN   
4     POLYGON ((130.84491 -12.41539, 130.84511 -12.4...     0820   
...                                                 ...      ...   
2664  GEOMETRYCOLLECTION (POLYGON ((146.80531 -41.30...     7268   
2665  GEOMETRYCOLLECTION (POLYGON ((146.80531 -41.30...     7270   
2666  GEOMETRYCOLLECTION (POLYGON ((146.80531 -41.30...     7275   
2667                                               None     7276   
2668                                               None     7277   

      SA2_MAINCODE_2011                SA2_NAME_2011  
0           701011002.0                  Darwin City  
1           701021025.0                   Nightcliff  
2           701021019.0                       Karama  
3                   NaN                          NaN  
4           701011008.0                  Stuart Park  
...                 ...                          ...  
2664        602031058.0           Dilston - Lilydale  
2665        602021053.0  Beauty Point - Beaconsfield  
2666        602021055.0         Grindelwald - Lanena  
2667        602021055.0         Grindelwald - Lanena  
2668        602011039.0                       Legana  

[2669 rows x 8 columns]

All not allocated POAs were found using the 2011 data.

In [43]:
# impute sa2 name and maincode
poa_w_sa2['sa2_maincode_2016'].fillna(poa_w_sa2['SA2_MAINCODE_2011'], 
                                      inplace=True)
poa_w_sa2['sa2_name_2016'].fillna(poa_w_sa2['SA2_NAME_2011'], inplace=True)

In [46]:
#impute sa2 geometry
poa_w_sa2 = poa_w_sa2.merge(sa2_bound[["sa2_maincode_2016", "geometry"]], 
                            on="sa2_maincode_2016", how="left")

In [48]:
poa_w_sa2 = poa_w_sa2[["poa_code_2016", "poa_name_2016", "sa2_maincode_2016", 
                       "sa2_name_2016", "geometry"]]

In [49]:
poa_w_sa2.isna().sum()

poa_code_2016        0
poa_name_2016        0
sa2_maincode_2016    0
sa2_name_2016        0
geometry             5
dtype: int64

In [50]:
poa_w_sa2[poa_w_sa2.geometry.isnull()]

poa_code_2016 poa_name_2016  sa2_maincode_2016  \
335            2230          2230        128011528.0   
2492           6530          6530        508051210.0   
2496           6537          6537        508021197.0   
2524           6707          6707        508021197.0   
2528           6713          6713        508061223.0   

                      sa2_name_2016 geometry  
335   Cronulla - Kurnell - Bundeena     None  
2492                      Geraldton     None  
2496                        Exmouth     None  
2524                        Exmouth     None  
2528                      Roebourne     None

In [52]:
# save final poa to sa2 table
poa_w_sa2.to_csv('../data/curated/poa_w_sa2.csv', index=False)